lib/OpenQA/Worker.pm

```diff
@@ -129,28 +336,290 @@ sub prepare_cache_directory {
             });
     }
-    else {
-        Mojo::IOLoop->stop;
+    # FIXME: Avoid so much elsif like in CommandHandler.pm.
+}
+
+sub _handle_job_status_changed {
+    my ($self, $job, $event_data) = @_;
+
+    my $job_id   = $job->id   // '?';
+    my $job_name = $job->name // '?';
+    my $client   = $job->client;
+    my $webui_host  = $client->webui_host;
+    my $status      = $event_data->{status};
+    my $reason      = $event_data->{reason};
+    my $current_job = $self->current_job;
+    if (!$current_job || $job != $current_job) {
+        die "Received job status update for job $job_id ($status) which is not the current one.";
     }
+
+    if ($status eq 'accepting') {
+        log_debug("Accepting job $job_id from $webui_host.");
+    }
+    elsif ($status eq 'accepted') {
+        $job->start();
+    }
+    elsif ($status eq 'setup') {
+        log_debug("Setting job $job_id from $webui_host up");
+    }
+    elsif ($status eq 'running') {
+        log_debug("Running job $job_id from $webui_host: $job_name.");
+    }
+    elsif ($status eq 'stopping') {
+        log_debug("Stopping job $job_id from $webui_host: $job_name - reason: $reason");
+    }
+    elsif ($status eq 'stopped') {
+        if (my $error_message = $event_data->{error_message}) {
+            log_error($error_message);
+        }
+        log_debug("Job $job_id from $webui_host finished - reason: $reason");
+        $self->current_job(undef);
+        $self->current_webui_host(undef);
+
+        # handle case when the worker should not continue to run e.g. because the user stopped it or
+        # a critical error occurred
+        if ($self->{_shall_terminate}) {
+            return $self->stop;
+        }
+
+        log_debug('Cleaning up for next job');
+        $self->_clean_pool_directory unless ($self->no_cleanup);
+    }
+    # FIXME: Avoid so much elsif like in CommandHandler.pm.
 }
 
-sub catch_exit {
-    my ($sig) = @_;
-    log_info("quit due to signal $sig");
-    stop('quit');
+sub _setup_pool_directory {
+    my ($self) = @_;
+
+    # skip if we have already locked the pool directory
+    my $pool_directory_fd = $self->{_pool_directory_lock_fd};
+    if (defined $pool_directory_fd) {
+        return $pool_directory_fd;
+    }
+
+    my $pool_directory = $self->pool_directory;
+    if (!$pool_directory) {
+        $self->current_error('No pool directory assigned.');
+        return undef;
+    }
+
+    try {
+        $pool_directory_fd = $self->_lock_pool_directory;
+        $self->_clean_pool_directory();
+    }
+    catch {
+        $self->current_error('Unable to lock and clean pool directory: ' . $_);
+    };
+    return $pool_directory_fd;
 }
 
-$SIG{HUP}  = \*catch_exit;
-$SIG{TERM} = \*catch_exit;
-$SIG{INT}  = \*catch_exit;
+sub _lock_pool_directory {
+    my ($self) = @_;
+
+    die 'no pool directory assigned' unless my $pool_directory = $self->pool_directory;
+
+    if (!-e $pool_directory) {
+        make_path($pool_directory);
+    }
```

Comment:
```perl
make_path($pool_directory) unless -e $pool_directory;
```

why?:

1. Saves two to three lines (if you also delete the above blank line)
2. States the intention about the desired end result in a more clean way: In the end one can be sure the pool directory is there, regardless if it existed in before or has to be created :)